src/main/java/org/folio/service/finance/transaction/EncumbranceRelationsHoldersBuilder.java

```diff
@@ -167,23 +162,28 @@ private void mapHoldersToTransactions(List<EncumbranceRelationsHolder> encumbran
   }
 
   public Future<List<EncumbranceRelationsHolder>> prepareEncumbranceRelationsHolder(CompositePurchaseOrder compPO,
-      CompositePurchaseOrder poFromStorage, RequestContext requestContext) {
+                                                                                    CompositePurchaseOrder poFromStorage, RequestContext requestContext) {
     List<EncumbranceRelationsHolder> holders = buildBaseHolders(compPO);
     return withFinances(holders, requestContext)
       .compose(v -> withExistingTransactions(holders, poFromStorage, requestContext));
   }
 
   public Future<Map<String, List<CompositePoLine>>> retrieveMapFiscalYearsWithCompPOLines(CompositePurchaseOrder compPO,
-      CompositePurchaseOrder poAndLinesFromStorage, RequestContext requestContext) {
+                                                                                          CompositePurchaseOrder poAndLinesFromStorage,
+                                                                                          RequestContext requestContext) {
     return prepareEncumbranceRelationsHolder(compPO, poAndLinesFromStorage, requestContext)
-      .map(erhList ->
-        erhList
-          .stream()
-          .filter(erh -> Objects.nonNull(erh.getCurrentFiscalYearId()))
-          .collect(groupingBy(
-            EncumbranceRelationsHolder::getCurrentFiscalYearId,
-            mapping(EncumbranceRelationsHolder::getPoLine, toList())
-          ))
+      .map(erhList -> erhList.stream()
+        .filter(erh -> erh.getCurrentFiscalYearId() != null)
+        .collect(Collectors.groupingBy(
+          EncumbranceRelationsHolder::getCurrentFiscalYearId,
+          Collectors.mapping(
+            EncumbranceRelationsHolder::getPoLine,
+            Collectors.collectingAndThen(
+              Collectors.toList(),
+              list -> list.stream().distinct().toList()
```

Comment:
so in this stream we jsut filter and prevent having duplicates in list by using distinct.
Probably without steams here would be easier to read but as you prefer.
Approving and also please add karate test